README.md
# Libftest
<img align="right" src="./srcs/img_readme.png" width="50%"/>
This repository contains script to test your libft project.
The script will do the following tests :

- Check if libft.h exists
- Check norme of libft.h
- Check if the Makefile file exists
- Check rules in Makefile
- Check compilation's flag in Makefile
- Compil library
- Check norme errors
- Check forbidden functions
- Tests functions

At the end of the tests, a deepthought file will be created, inside which you
can find all the results and error/failure messages. You can also see your
results in tests directory.

The script will compil your files with your libft, so you must have your
Makefile up to date.

:warning:All the tests made are not the official tests:warning:

## Getting Started

### Installation

```bash
git clone https://github.com/jtoty/Libftest
```
The project needs `clang` compiler. Make sure it is installed.

#### Old libft subject
To test the old libft subject, you have to go to the branch [old-libft](https://github.com/jtoty/Libftest/tree/old-libft).

### Configuration

If you run the script for the first time, it will automatically create
`my_config.sh` file and ask you to edit it with the path of your libft
project.
You can also choose the colors that will be used and the path where the
deepthought file will be created.

## Running script

Run the script from the directory where you cloned
```bash
bash grademe.sh
```
or run the script with the path
```bash
bash /path/where/you/cloned/grademe.sh
```

### Options available
| Option | Description |
| --- | --- |
| `-h`<br />`--help` | Display help and exit |
| `-c` | Disable color |
| `-s` | Disable searching Makefile and author files |
| `-m` | When compiling library, test all the Makefile's rules (instead of doing only make re and checking if other rules exist). |
| `-l` | Disable compiling library |
| `-f` | Disable checking forbidden functions |
| `-n` | Disable norminette |
| `-u` | Disable checking updates at launch |
| `-p1` | Disable part 1 tests |
| `-p2` | Disable part 2 tests |
| `-b`| Disable bonus part tests |
| `-a` | Disable additional part tests |
| `-op1`| Do only part 1 tests |
| `-op2`| Do only part 2 tests |
| `-ob` | Do only bonus part tests |
| `-oa` | Do only additional part tests |
| `ft_function` | Test only this function |

Add successively all options you want, in the order you want.
For example :
```bash
bash grademe.sh ft_atoi -f ft_strlen -n
```

### Supported functions
All the supported functions are listed [on this page](https://github.com/jtoty/Libftest/blob/master/supported_functions.md).
## Contribution

Any suggestions or bugs reporting ?
Contact jtoty@student.42.fr

### Contributors
- jmichaud : https://github.com/MrJe
- tlernoul


changelog
~~~~~~~~~~~~~~~~~~~~~~~~~~2021/03/05~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Missing casts on ft_lst functions

~~~~~~~~~~~~~~~~~~~~~~~~~~2021/12/01~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Compilation error on ft_memchr

~~~~~~~~~~~~~~~~~~~~~~~~~~2020/23/05~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Pointer compare warning on ft_split

~~~~~~~~~~~~~~~~~~~~~~~~~~2020/23/03~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- False positive test #2 in ft_memccpy

~~~~~~~~~~~~~~~~~~~~~~~~~~2019/12/04~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Norm check
- Abort in ft_lstclear
- Abort in ft_lstdelone
- Test in ft_memmove

~~~~~~~~~~~~~~~~~~~~~~~~~~2019/11/30~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Works with the new libft subject

~~~~~~~~~~~~~~~~~~~~~~~~~~2019/01/17~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Error in ft_striteri
- Error in ft_strmapi

~~~~~~~~~~~~~~~~~~~~~~~~~~2018/06/29~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strtok

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/11/16~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strlcat

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/11/13~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Error in ft_strlcpy
- Error in ft_lstdelone

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/09/25~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strcasestr

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/04/14~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strncasecmp
- One test for ft_strrchr

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/04/02~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Legend at the end of the script

#Fixed
- Tests for fd functions

#Changed
- Option -q becomes -m and allows to test all the Makefile's
rules instead of doing only make re

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/28~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Abort detection
- Bus error detection
- Timeout detection

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/25~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Display format of -h/--help option

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Test 1 of ft_strnequ

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/01/23~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Option -d : allows to perform the tests even if the files
are in directories

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/22~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strsub

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/20~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_itoa
- One test for ft_itoa_base

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/19~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_itoa_base

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/16~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strchr
- One test for ft_strrchr

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Option -op1 : do only part 1 tests
- Option -op2 : do only part 2 tests
- Option -ob : do only bonus part tests
- Option -oa : do only additional part tests
- Tests for ft_lstadd
- Tests for ft_lstdelone
- Tests for ft_lstdel

#Changed
- Tests of ft_strchr
- Tests of ft_strrchr
- Tests of ft_strmap
- Tests of ft_strmapi
- Tests of ft_striter
- Tests of ft_striteri

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/04~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_memmove

#Changed
- Test 1 of ft_strcat
- Test 3 of ft_strcat
- Test 1 of ft_strncat
- Test 2 of ft_strncat
- Test 5 of ft_strncat
- Test 1 of ft_strlcat
- Test 3 of ft_strlcat
- Test 4 of ft_strlcat
- Test 5 of ft_strlcat
- Test 6 of ft_strlcat
- Test 8 of ft_strlcat

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/03~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strncpy

#Changed
- Image for README.md

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/30~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- New character when display tests

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/30~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strlcat
- One test for ft_bzero

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/29~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Display error in deepthought file

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/21~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Forbidden functions detection. nm command is used now for
better detection. Option -f to disable it
- List of forbidden functions in part 1
- Improve rapidity when checking norme

#Fixed
- Bug when checking forbidden functions

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/21~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Bug when checking forbidden functions

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/20~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strcasecmp

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/19~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Test 2 of ft_strnequ (previous test was wrong)
- Test 3 of ft_strnequ (previous test was wrong)
- Test 5 of ft_strnequ (previous test was wrong)
- Test 6 of ft_strnequ (previous test was wrong)

#Added
- Option -h / --help : display help and exit
- One additional function : ft_strndup

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Remove
- Tests for ft_lstadd, ft_lstdel & ft_lstdelone due to errors

#Added
- One test for ft_strlcat
- Display name of each part in deepthought file

#Changed
- Test 13 of ft_atoi
- Test 2 of ft_atoi
- Test 18 of ft_atoi

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Tests for ft_lstadd
- One test for ft_strnstr
- One test for ft_strlcat
- Tests for ft_lstdel
- Tests for ft_lstdelone
- One test for ft_strlcpy
- ft_*.o are now removed at the end of the tests
- One test for ft_strnew

#Changed
- Test 5 of ft_strlcat
- Test 1 of ft_strnew

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Tests for ft_strnequ

#Added
- Tests for ft_lstnew
- Tests for ft_lstiter
- Tests for ft_lstmap

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Additional test for ft_strlcat
- Additional test for ft_strcmp
- Additional test for ft_strtrim
- Additional test for ft_strncmp
- Additional test for ft_strsplit

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/05~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Additional test for ft_memmove

#Changed
- One test of ft_memcmp
_ One test of ft_strnew

#Fixed
- Checking content of author file


srcs/variables.sh
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    variables.sh                                       :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jtoty <jtoty@student.42.fr>                +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2017/01/23 18:27:13 by jtoty             #+#    #+#              #
#    Updated: 2018/06/29 13:28:44 by jtoty            ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

#!/bin/bash

###################################################
#                 Part1 functions                 #
###################################################

Part1_func=(
	'ft_atoi.c'		\
	'ft_bzero.c'	\
	'ft_calloc.c'	\
	'ft_isalnum.c'	\
	'ft_isalpha.c'	\
	'ft_isascii.c'	\
	'ft_isdigit.c'	\
	'ft_isprint.c'	\
	'ft_memccpy.c'	\
	'ft_memchr.c'	\
	'ft_memcmp.c'	\
	'ft_memcpy.c'	\
	'ft_memmove.c'	\
	'ft_memset.c'	\
	'ft_strchr.c' 	\
	'ft_strdup.c'	\
	'ft_strlcat.c'	\
	'ft_strlcpy.c'	\
	'ft_strlen.c'	\
	'ft_strncmp.c'	\
	'ft_strnstr.c'	\
	'ft_strrchr.c'	\
	'ft_tolower.c'	\
	'ft_toupper.c'	\
)

Part1_func_authorized=(
	'0'	\
	'0'	\
	'1'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'1'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
)

Part1_func_activation=(
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
)

###################################################
#                 Part2 functions                 #
###################################################

Part2_func=(
	'ft_itoa.c'			\
	'ft_putchar_fd.c'	\
	'ft_putendl_fd.c'	\
	'ft_putnbr_fd.c'	\
	'ft_putstr_fd.c'	\
	'ft_split.c'		\
	'ft_strjoin.c'		\
	'ft_strmapi.c'		\
	'ft_strtrim.c'		\
	'ft_substr.c'		\
)

Part2_func_authorized=(
	'1'	\
	'4'	\
	'4'	\
	'4'	\
	'4'	\
	'3'	\
	'1'	\
	'1'	\
	'1'	\
	'1'	\
)

Part2_func_activation=(
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
)

###################################################
#                 Bonus functions                 #
###################################################

Bonus_func=(
	'ft_lstadd_back.c'	\
	'ft_lstadd_front.c'	\
	'ft_lstclear.c'		\
	'ft_lstdelone.c'	\
	'ft_lstiter.c'		\
	'ft_lstlast.c'		\
	'ft_lstmap.c'		\
	'ft_lstnew.c'		\
	'ft_lstsize.c'		\
)

Bonus_func_authorized=(
	'0'	\
	'0'	\
	'2'	\
	'2'	\
	'0'	\
	'0'	\
	'3'	\
	'3'	\
	'0'	\
)

Bonus_func_activation=(
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
)

###################################################
#              Additional functions               #
###################################################

Additional_func=(
	'ft_isblank.c'			\
	'ft_iscntrl.c'			\
	'ft_isgraph.c'			\
	'ft_islower.c'			\
	'ft_isspace.c'			\
	'ft_isupper.c'			\
	'ft_isxdigit.c'			\
	'ft_itoa_base.c'		\
	'ft_lstadd.c'			\
	'ft_lstdel.c'			\
	'ft_memalloc.c'			\
	'ft_memdel.c'			\
	'ft_putchar.c'			\
	'ft_putendl.c'			\
	'ft_putnbr.c'			\
	'ft_putstr.c'			\
	'ft_str_is_alpha.c'		\
	'ft_str_is_lowercase.c'	\
	'ft_str_is_numeric.c'	\
	'ft_str_is_printable.c'	\
	'ft_str_is_uppercase.c'	\
	'ft_strcapitalize.c'	\
	'ft_strcasecmp.c'		\
	'ft_strcasestr.c'		\
	'ft_strcat.c'			\
	'ft_strclr.c'			\
	'ft_strcmp.c'			\
	'ft_strcpy.c'			\
	'ft_strdel.c'			\
	'ft_strequ.c'			\
	'ft_striter.c'			\
	'ft_striteri.c'			\
	'ft_strlowcase.c'		\
	'ft_strmap.c'			\
	'ft_strncasecmp.c'		\
	'ft_strncat.c'			\
	'ft_strncpy.c'			\
	'ft_strndup.c'			\
	'ft_strnequ.c'			\
	'ft_strnew.c'			\
	'ft_strstr.c'			\
	'ft_strtok.c'			\
	'ft_strupcase.c'		\
)

Additional_func_authorized=(
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'1'	\
	'0'	\
	'2'	\
	'1'	\
	'2'	\
	'4'	\
	'4'	\
	'4'	\
	'4'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'2'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'1'	\
	'0'	\
	'0'	\
	'0'	\
	'1'	\
	'0'	\
	'1'	\
	'0'	\
	'0'	\
	'0'	\
)

Additional_func_activation=(
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
	'0'	\
)

###################################################
#                 Other variables                 #
###################################################

tab_all_part=('Part1_func' 'Part2_func' 'Bonus_func' 'Additional_func')

num_sys_func=('1' '2' '4')
system_func=('void' 'malloc' 'free' 'printf' 'write')

NORME_COL=23
CHEAT_COL=50
COMPIL_COL=38
TEST_COL=67
RESULT_COL=87
TITLE_LENGTH=92
CHAR_LENGTH="-"
CHAR_WIDTH="|"

DIRECTORY=0
OPT_NO_LIBRARY=0
OPT_FULL_MAKEFILE=0
OPT_NO_SEARCH=0
OPT_NO_COLOR=0
OPT_NO_FORBIDDEN=0
OPT_NO_NORMINETTE=0
OPT_NO_UPDATE=0
OPT_NO_PART1=0
OPT_NO_PART2=0
OPT_NO_BONUS=0
OPT_NO_ADDITIONAL=0
ACTIVATE_PART1=0
ACTIVATE_PART2=0
ACTIVATE_BONUS=0
ACTIVATE_ADDITIONAL=0
CHECK_IN_PART1=1
CHECK_IN_PART2=1
CHECK_IN_BONUS=1
CHECK_IN_ADDITIONAL=1
